pyteal/ast/abi/transaction.py

```diff
@@ -0,0 +1,196 @@
+from abc import abstractmethod
+from typing import TypeVar, Union, cast
+from pyteal.ast.abi.type import BaseType, ComputedValue, TypeSpec
+from pyteal.ast.expr import Expr
+from pyteal.ast.int import Int
+from pyteal.ast.txn import TxnType
+from pyteal.types import TealType
+from pyteal.errors import TealInputError
+
+T = TypeVar("T", bound=BaseType)
+
+
+class TransactionTypeSpec(TypeSpec):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def new_instance(self) -> "Transaction":
+        return Transaction(self)
+
+    def annotation_type(self) -> "type[Transaction]":
+        return Transaction
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def byte_length_static(self) -> int:
+        raise TealInputError("Transaction Types don't have a static size")
+
+    def storage_type(self) -> TealType:
+        return TealType.uint64
+
+    def __eq__(self, other: object) -> bool:
+        return type(self) is type(other)
+
+    @abstractmethod
+    def __str__(self) -> str:
+        return TxnType.Any.name
+
+
+TransactionTypeSpec.__module__ = "pyteal"
+
+
+class Transaction(BaseType):
+    def __init__(self, spec: TransactionTypeSpec) -> None:
+        super().__init__(spec)
+
+    def type_spec(self) -> TransactionTypeSpec:
+        return cast(TransactionTypeSpec, super().type_spec())
+
+    def get(self) -> Expr:
```

Comment:
Suggestion: don't make `get` return the index/offset (I can't tell which the implementation uses yet) -- most people won't care about this. Make `get` return the correct `TxnObject`, e.g. `return Gtxn[<index here>]`. Then the caller can access all the fields on it.

If you want to return the index, a separate `index` method seems more appropriate.

Reply:
I agree that `get` is better to return some TxnObject since that is how it will be used. 

I think it might be a departure from the other sort of reference types since they offer both `get` for the raw application_arg decoded vs `deref` for dereferencing the reference type. 

As long as this is ok, I'm happy to make this `get` return the TxnObject but I wonder if I should go back to the reference types and make them return the deref by default and provide a different method to get the app arg? 

@ahangsu @tzaffi any thoughts?

Reply:
I think we should do is aim for consistency. @jasonpaulos argues that most developers will prefer the actual `TxnObject` over its offset. A similar argument could be made for reference types such as `Asset` etc. If we generally expect `get` to give the "raw" data for an abi-type, then we should continue with Ben's original approach here but communicate in our docs as clearly as possible that this is the purpose of `get()`. But my hunch is that Jason's assessment of usage patterns is spot-on, and that `get()` (and similarly `__get_item__()`) should act in the most convenient manner. 

If we all agree with this philosophy around `get`, I recommend adding to this PR changes that make `deref()` become `get()` and then also allow users to retrieve the raw information either with something like `index()`as Jason suggests, or a variant such as `array_index()` etc. (I prefer `index` even if it clashes with usages of `index()` in other types because it makes sense to me to have polymorphic versions of `index()` based on the type's context).

Reply:
I've tweaked both ref types and txn types here to return what most people would probably expect